src/gadgets/scalar/range.rs

```diff
@@ -0,0 +1,249 @@
+            .iter()
+            .enumerate()
+            .fold(BlsScalar::zero(), |scalar_accum, (idx, mut bit)| {
+                if idx >= num_bits as usize {
+                    bit = &0u8;
+                };
+                let bit_var = composer.add_input(BlsScalar::from(*bit as u64));
+                // Apply boolean constraint to the bit.
+                composer.boolean_gate(bit_var);
+                // Accumulate the bit multiplied by 2^(i-1) as a variable
+                var_accumulator = composer.add(
+                    (BlsScalar::one(), var_accumulator),
+                    (BlsScalar::from(2u64).pow(&[idx as u64, 0, 0, 0]), bit_var),
+                    BlsScalar::zero(),
+                    BlsScalar::zero(),
+                );
+                scalar_accum
+                    + (BlsScalar::from(*bit as u64)
+                        * BlsScalar::from(2u64).pow(&[idx as u64, 0, 0, 0]))
+            });
+    // Compute `Chi(x)` =  Sum(vi * 2^(i-1)) - (x + b').
+    let witness_plus_b_prime = composer.add_input(witness + b_prime);
+    // Note that the result will be equal to: `0 (if the reangeproof holds)
+    // or any other value if it doesn't.
+    let chi_x_var = composer.add(
+        (BlsScalar::one(), witness_plus_b_prime),
+        (-BlsScalar::one(), var_accumulator),
+        BlsScalar::zero(),
+        BlsScalar::zero(),
+    );
+    // It is possible to replace a constraint $\chi(\mathbf{x})=0$ on variables
+    // $\mathbf{x}$ with a set of constraints $\psi$ on  new variables
+    // $(u,y,z)$ such that $y=1$ if $\chi$ holds and $y=0$ otherwise.
+    // We introduce new variables $u,y,z$ that are computed as follows:
+    //
+    // u &= \chi(\mathbf{x});\\
+    // y &=\begin{cases}
+    // 0,& \text{if }u\neq 0;\\
+    // 1,& \text{if }u=0.
+    // \end{cases}\\
+    // z&=\begin{cases}
+    // 1/u,& \text{if }u\neq 0;\\
+    // 0,& \text{if }u=0.
+    // \end{cases}
+    let u = witness + b_prime - accumulator;
+    // Conditionally assign `1` or `0` to `y`.
+    let y = if u == BlsScalar::zero() {
+        composer.add_input(BlsScalar::one())
+    } else {
+        composer.add_input(BlsScalar::zero())
+    };
+    // Conditionally assign `1/u` or `0` to z
+    let mut z = zero;
+    if u != BlsScalar::zero() {
+        // If u != zero -> `z = 1/u`
+        // Otherways, `u = 0` as it was defined avobe.
+        // Check inverse existance, otherways, err.
+        if u.invert().is_none().into() {
+            return Err(GadgetErrors::NonExistingInverse.into());
+        };
+        // Safe to unwrap here.
+        z = composer.add_input(u.invert().unwrap());
+    }
+    // We can safely unwrap `u` now.
+    // Now we need to check the following to ensure we can provide a boolean
+    // result representing wether the rangeproof holds or not:
+    // `u = Chi(x)`.
+    // `u * z = 1 - y`.
+    // `y * u = 0`.
+    let one = composer.add_input(BlsScalar::one());
+    composer.add_gate(
```

Comment:
I don't understand why we use an `add_gate` instead of `constrain_to_constant` since we are stating `chi_x_var (variable) == u (scalar)`

Reply:
Constraining to a constant will make the circuits that use this valid just for one value of `Chi`. We don't want to set any of these values to be printed inside of the circuit. So that the gadget/circuit is reusable.